include/RAJA/policy/tbb/reduce.hpp

```diff
@@ -0,0 +1,407 @@
+//   this list of conditions and the disclaimer (as noted below) in the
+//   documentation and/or other materials provided with the distribution.
+//
+// * Neither the name of the LLNS/LLNL nor the names of its contributors may
+//   be used to endorse or promote products derived from this software without
+//   specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL LAWRENCE LIVERMORE NATIONAL SECURITY,
+// LLC, THE U.S. DEPARTMENT OF ENERGY OR CONTRIBUTORS BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "RAJA/internal/MemUtils_CPU.hpp"
+#include "RAJA/pattern/reduce.hpp"
+#include "RAJA/policy/tbb/policy.hpp"
+#include "RAJA/util/types.hpp"
+
+#include <tbb/tbb.h>
+#include <memory>
+#include <tuple>
+
+namespace RAJA
+{
+
+namespace detail
+{
+template <typename T, typename Op>
+class ReduceTBB
+{
+public:
+  using Reduce = Op;
+  using value_type = T;
+
+  //! TBB native per-thread container
+  std::shared_ptr<tbb::combinable<T>> data;
+
+  //! prohibit compiler-generated default ctor
+  ReduceTBB() = delete;
+
+  //! prohibit compiler-generated copy assignment
+  ReduceTBB &operator=(const ReduceTBB &) = delete;
+
+  //! compiler-generated copy constructor
+  ReduceTBB(const ReduceTBB &) = default;
+
+  //! compiler-generated move constructor
+  ReduceTBB(ReduceTBB &&) = default;
+
+  //! compiler-generated move assignment
+  ReduceTBB &operator=(ReduceTBB &&) = default;
+
+  //! constructor requires a default value for the reducer
+  explicit ReduceTBB(T init_val, T initializer = T())
+      : data(
+            std::make_shared<tbb::combinable<T>>([=]() { return initializer; }))
+  {
+    data->local() = init_val;
+  }
+
+  /*!
+   *  \return the calculated reduced value
+   */
+  operator T() const { return data->combine(Op{}); }
+
+  /*!
+   *  \return the calculated reduced value
+   */
+  T get() const { return operator T(); }
+
+protected:
+  /*!
+   *  \return update the local value
+   */
+  void combine(const T &other) { data->local() = Op{}(data->local(), other); }
+  /*!
+   *  \return update the local value
+   */
+  void combine(const T &other) const
+  {
+    data->local() = Op{}(data->local(), other);
+  }
+};
```

Comment:
Can we add a blank line here and some below to separate classes/structs and make it easier to visually parse?